docs/adr/0009-async-rust.md

```diff
@@ -0,0 +1,151 @@
+# Using Async Rust
+
+* Status: draft
+* Deciders: 
+* Date: ???
+
+## Context and Problem Statement
+
+Our Rust components are currently written as using synchronous Rust.
+The components are then wrapped in Kotlin to present an async interface.
+Swift also wraps them to present an async-style interface, although it currently uses `DispatchQueue` and completion handlers rather than `async` functions.
+
+UniFFI has been adding async capabilities in the last year and it seems possible to switch to using async Rust and not having an async wrapper.
+It also seems possible to auto-generate the async wrapper with UniFFI.
+
+What should our async strategy be?
+
+### Scope
+
+This ADR discusses what our general policy on wrapper code should be.
+It does not cover how we should plan our work.
+If we decide to embrace async Rust, we do not need to commit to any particular timeline for actually switching to it.
+
+### Desktop and gecko-js
+
+On desktop, we can't write async wrappers because it's not possible in Javascript.
+Instead we use a strategy where every function is automatically wrapped as async in the C++ layer.
+Using a config file, it's possible to opt-out of this strategy for particular functions/methods.
+
+### Android-components
+
+In Kotlin, the async wrapper layer currently lives in `android-components`.
+For the purposes of this ADR, it doesn't really matter, and this ADR will not make a distinction between wrapper code in our repo and `android-components`.
```

Comment:
At first, I was wondering if this was a bit inconsistent with your "Wrapper code" ADR, because that ADR (seems to?) explicitly exclude wrappers outside of Application Services.

But thinking about it a bit more, the way you've written this makes sense to me! All our wrappers—wherever they live—do two things:

1. Wrap the synchronous Rust component interface in an asynchronous foreign interface.
2. Bridge the Rust interface to platform-specific Android and iOS interfaces (`WorkManager`, `BGTaskScheduler`, `UIKit`, etc.)

We can lift (1) into Application Services, and still keep the wrappers in our consuming applications for (2).